yacl/crypto/primitives/tpre/tpre_test.cc

```diff
@@ -0,0 +1,119 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "yacl/crypto/primitives/tpre/tpre.h"
+
+#include "absl/strings/escaping.h"
+#include "gtest/gtest.h"
+
+namespace yacl::crypto::test {
+
+TEST(TpreTest, Test1) {
+  std::string iv = "123456789";
+  TPRE tpre;
+  Keys keys;
+
+  /************************* Phase 1 *************************/
+  // Start testing encryption and decryption functions
+  std::unique_ptr<EcGroup> ecc_group = EcGroupFactory::Create("sm2");
+
+  std::pair<Keys::PublicKey, Keys::PrivateKey> key_pair_A =
+      keys.GenerateKeyPair(ecc_group);
+
+  std::unique_ptr<Keys::PublicKey> pk_A(new Keys::PublicKey(key_pair_A.first));
+  std::unique_ptr<Keys::PrivateKey> sk_A(
+      new Keys::PrivateKey(key_pair_A.second));
+
+  // test tpre.Encrypt
+  std::string message = "hellooooooooooooo, I'am 63, who are you?";
+
+  std::pair<Capsule::CapsuleStruct, std::vector<uint8_t>> ct_1 =
+      tpre.Encrypt(ecc_group, pk_A, iv, message);
+
+  // test tpre.Decrypt
+
+  std::unique_ptr<Capsule::CapsuleStruct> capsule_struct_1(
+      new Capsule::CapsuleStruct(ct_1.first));
+  std::string message_1 =
+      tpre.Decrypt(ecc_group, capsule_struct_1, iv, ct_1.second, sk_A);
+
+  // Determine if decryption was successful
+  EXPECT_EQ(message, message_1);
+
+  // End testing encryption and decryption functions
+
+  /************************* Phase 2 *************************/
+  // Start testing encryption, re-encryption, and decryption functions
+
+  // Second test tpre.Encrypt
+  std::string message_2 =
+      "If you were a teardrop;In my eye, For fear of losing you, I would never "
+      "cry. And if the golden sun, Should cease to shine its light, Just one "
+      "smile from you, Would make my whole world bright.";
+
+  std::pair<Capsule::CapsuleStruct, std::vector<uint8_t>> ct_2 =
+      tpre.Encrypt(ecc_group, pk_A, iv, message_2);
+
+  // test keys->GenerateReKey
+  std::pair<Keys::PublicKey, Keys::PrivateKey> key_pair_B =
+      keys.GenerateKeyPair(ecc_group);
+  std::unique_ptr<Keys::PublicKey> pk_B(new Keys::PublicKey(key_pair_B.first));
+  std::unique_ptr<Keys::PrivateKey> sk_B(
+      new Keys::PrivateKey(key_pair_B.second));
+
+  int N = 5;  // Number of all participants
+  int t = 4;  // Threshold
+
+  std::vector<Keys::KFrag> kfrags =
+      keys.GenerateReKey(ecc_group, sk_A, pk_A, pk_B, N, t);
+
+  // test tpre.ReEncrypt
+  std::pair<std::vector<std::unique_ptr<Capsule::CFrag>>, std::vector<uint8_t>>
+      re_ct_set;
+
+  // You need to meet the number of participants to successfully decrypt,
+  // otherwise decryption will not be successful
+
+  for (int i = 0; i < t; i++) {
+    std::unique_ptr<Capsule::CapsuleStruct> capsule_struct_2(
+        new Capsule::CapsuleStruct(ct_2.first));
+    std::pair<std::unique_ptr<Capsule::CapsuleStruct>, std::vector<uint8_t>>
+        ct_2_i = {std::move(capsule_struct_2), ct_2.second};
```

Comment:
These `std::move` after could also remove, if we just use `Capsule::CapsuleStruct` instead of wrapper it as a smart pointer `std::unique_ptr<Capsule::CapsuleStruct>`.